File: python/ch03_agent/agent_ch03/core/profiles.py
"""章节能力白名单。第 3 章只开放 loop 和 powershell。

Java 对照：ChapterProfile 类似一个不可变配置 DTO，P01 类似 public static final 常量。
它防止第 3 章意外装配后续章节才应该出现的能力。
"""

from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class ChapterProfile:
    """某个章节允许启用的能力集合。"""

    chapter: int  # 章节编号，例如 1。
    capabilities: frozenset[str]  # 不可变集合，类似 Collections.unmodifiableSet。


P01 = ChapterProfile(1, frozenset({"loop", "powershell"}))  # 第一章固定能力配置。
P02 = ChapterProfile(2, frozenset({"loop", "powershell", "tool_registry", "files"}))
P03 = ChapterProfile(3, frozenset({"loop", "powershell", "tool_registry", "files", "policy"}))


def profile_for_chapter(chapter: int) -> ChapterProfile:
    """根据章节号取得配置；未迁移的章节明确报错。"""
    if chapter == 1:
        return P01
    if chapter == 2:
        return P02
    if chapter == 3:
        return P03
    raise ValueError(f"第 {chapter} 章尚未迁移为 Python 版本")

File: typescript/chapters/ch04/src/core/profiles.ts
// P04 将 hooks 加入章节能力集，允许在固定生命周期点执行受限扩展。
// 章节能力快照：每章一个冻结单例，组合根用引用相等校验 profile 来源。
export type Capability = "loop" | "powershell" | "tool_registry" | "files" | "policy" | "hooks";

export interface ChapterProfile {
  readonly chapter: number;
  readonly capabilities: ReadonlySet<Capability>;
}

class CapabilitySet implements ReadonlySet<Capability> {
  // 用不可导出底层 Set 的包装实现 ReadonlySet，防止 profile 能力被外部修改。
  readonly #values: Set<Capability>;

  constructor(values: readonly Capability[]) {
    this.#values = new Set(values);
  }

  get size(): number {
    return this.#values.size;
  }

  has(value: Capability): boolean {
    return this.#values.has(value);
  }

  entries(): SetIterator<[Capability, Capability]> {
    return this.#values.entries();
  }

  keys(): SetIterator<Capability> {
    return this.#values.keys();
  }

  values(): SetIterator<Capability> {
    return this.#values.values();
  }

  forEach(
    callbackfn: (value: Capability, value2: Capability, set: ReadonlySet<Capability>) => void,
    thisArg?: unknown,
  ): void {
    this.#values.forEach((value) => {
      callbackfn.call(thisArg, value, value, this);
    });
  }

  [Symbol.iterator](): SetIterator<Capability> {
    return this.values();
  }
}

export const P01: ChapterProfile = Object.freeze({
  chapter: 1,
  capabilities: new CapabilitySet(["loop", "powershell"]),
});

export const P02: ChapterProfile = Object.freeze({
  chapter: 2,
  capabilities: new CapabilitySet(["loop", "powershell", "tool_registry", "files"]),
});

export const P03: ChapterProfile = Object.freeze({
  chapter: 3,
  capabilities: new CapabilitySet(["loop", "powershell", "tool_registry", "files", "policy"]),
});

export const P04: ChapterProfile = Object.freeze({
  chapter: 4,
  capabilities: new CapabilitySet([
    "loop",
    "powershell",
    "tool_registry",
    "files",
    "policy",
    "hooks",
  ]),
});

export function profileForChapter(chapter: number): ChapterProfile {
  // 仅返回模块内冻结的单例，供组合根用引用相等性校验 profile 来源。
  if (chapter === 1) {
    return P01;
  }
  if (chapter === 2) {
    return P02;
  }
  if (chapter === 3) {
    return P03;
  }
  if (chapter === 4) {
    return P04;
  }
  if (!Number.isInteger(chapter) || chapter < 1 || chapter > 20) {
    throw new Error("chapter must be an integer from 1 to 20");
  }
  throw new Error(`Chapter ${chapter} has not been migrated to TypeScript yet`);
}
